drivers: sensor: adxl375 accelerometer driver support

AarC10 opened the pull request:
Basic support for an ADXL375 accelerometer driver similar to other ADXL37X models over I2C and SPI tested on an Adafruit ADXL375 breakout board. Support for configuring output data rate and low power mode, as well as retrieving acceleration values.

Signed-off-by: Aaron Chan [aarchan108@gmail.com](mailto:aarchan108@gmail.com)

thedjnK commented:
See https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-message-guidelines and squash the same commits into one

AarC10 commented:
Hi @MaureenHelm Could I please get a review of this PR to gain a better understanding of what to fix and how to improve so this could get merged for my org. Thanks!

aasinclair commented on drivers/sensor/adxl375/adxl375.c:
```diff
@@ -0,0 +1,212 @@
+
+static void adxl375_accel_convert(struct sensor_value *result, int16_t sample_val)
+{
+	result->val1 = (sample_val * SENSOR_G * ADXL375_MG2G_MULTIPLIER) / 1000000;
+	result->val2 = ((int16_t)(sample_val * SENSOR_G * ADXL375_MG2G_MULTIPLIER)) % 1000000;
```
This cast will truncate the value.
I suggest creating an intermediate scaled value, e.g.:
```suggestion
	int64_t value = (int64_t)sample_val * SENSOR_G * ADXL375_MG2G_MULTIPLIER;

	result->val1 = value / 1000000;
	result->val2 = value % 1000000;
```

AarC10 replied:
Ooh very nice catch. Did not see that. Thank you

aasinclair commented on drivers/sensor/adxl375/adxl375.c:
```diff
@@ -0,0 +1,212 @@
+
+	int ret = data->hw_tf->read_reg_multiple(dev, ADXL375_DATAX0, buff, 6);
+
+	data->sample.x = (int16_t)((buff[1] << 8) | (buff[0]));
```
You can use `sys_get_le16` here

AarC10 replied:
Thank you

aasinclair commented on drivers/sensor/adxl375/adxl375.c:
```diff
@@ -0,0 +1,212 @@
+
+	// Need to force bits 3, 1 and 0 to set since the register defaults to 0
+	// See datasheet pages 20 and 24
+	ret = adxl375_set_data_format(dev, 0b00001011);
```
Please avoid binary literals

aasinclair commented on dts/bindings/sensor/adi,adxl375-common.yaml:
```diff
@@ -0,0 +1,56 @@
+include: sensor-device.yaml
+
+properties:
+  odr:
```
You could implement this as a string property and use DT_INST_ENUM_IDX.
That way the user can use the required value directly in their devicetree config, e.g:
https://github.com/zephyrproject-rtos/zephyr/blob/f0b975032a68c2da38c34f01b836edb5e9e62b12/dts/bindings/pinctrl/ambiq%2Capollo4-pinctrl.yaml#L76-L82

aasinclair reviewed (changes requested):
Thanks for this driver, it's looking really good.

aasinclair commented on drivers/sensor/adxl375/adxl375_i2c.c:
```diff
@@ -0,0 +1,91 @@
+
+	data->hw_tf = &adxl375_i2c_transfer_fn;
+
+	if (!device_is_ready(config->i2c.bus)) {
```
```suggestion
	if (!i2c_is_ready_dt(config->i2c)) {
```

MaureenHelm replied:
Doesn't look like this has been addressed

AarC10 replied:
Ah sorry about that. Must've slipped through this. Will be addressed when I try resolve all other comments and compliance checks.

aasinclair commented on drivers/sensor/adxl375/adxl375_i2c.c:
```diff
@@ -0,0 +1,91 @@
+
+static int adxl375_i2c_reg_read(const struct device *dev, uint8_t reg_addr, uint8_t *reg_data)
+{
+	return adxl375_bus_access(dev, ADXL375_REG_READ(reg_addr), reg_data, 1);
```
Couldn't you just use the i2c commands directly in these helper function?

AarC10 replied:
Agreed and I can make this change. Just as a heads up though that in the adxl372 and adxl367 code, they do something similar. I was trying my best to keep things consistent.

aasinclair replied:
Ah I see, at some point in the past the SPI driver was copied to make the I2C driver.
I don't think you need to keep consistency with the old driver, and it isn't worth making a special change to the older driver unless it is being updated for another reason.

AarC10 commented on drivers/sensor/adxl375/adxl375_i2c.c:
```diff
@@ -0,0 +1,84 @@
+{
+	const struct adxl375_dev_config *config = dev->config;
+
+	return i2c_burst_read_dt(&config->i2c, ADXL375_TO_I2C_REG(ADXL375_REG_READ(reg_addr)),
+				 reg_data, 1);
```
Yep guess that came from the SPI implementation as well. The macros in general don't seem to be needed for the I2C implementation, although the sensor still functions either way.

MaureenHelm commented on dts/bindings/sensor/adi,adxl375-i2c.yaml:
Where is the SPI version of this file?

MaureenHelm reviewed (changes requested):
Thank you for the contribution and apologies for the delay in my review.

Please add this sensor to `tests/drivers/build_all/sensor/` to ensure it builds in CI, and also address the failed compliance check.

@amiclaus please take a look

amiclaus commented on drivers/sensor/adxl375/adxl375.c:
```diff
@@ -0,0 +1,215 @@
+		LOG_ERR("Failed to put in standby mode");
+	}
+
+	return ret;
```
if you check throughout the function `ret != 0` then you can return 0 directly here.

amiclaus commented on drivers/sensor/adxl375/adxl375.h:
```diff
@@ -0,0 +1,348 @@
+/*
+ * ADXL375 registers definition
+ */
+#define ADXL375_DEVID            0x00u /* Analog Devices accelerometer ID */
```
i'd properly align all the macro values. improves readability.

AarC10 replied:
If you're talking about having it one space apart for each, I don't think I have power over this. I clang-formatted every file to meet the CI checks and this is how the formatter wanted to do it

amiclaus commented on drivers/sensor/adxl375/adxl375.h:
```diff
@@ -0,0 +1,348 @@
+	const struct adxl375_transfer_function *hw_tf;
+	struct adxl375_fifo_config fifo_config;
+#ifdef CONFIG_ADXL375_TRIGGER
+	struct gpio_callback gpio_cb;
```
trigger supports seems to be declared, but not implemented.

AarC10 replied:
In response to this comment and the one in the KConfig, no support for it yet. I wanted to get basic functionality for a driver in first and see about supporting it another time if I found time or if someone else was interested in doing so. I can remove this unused code.

amiclaus commented on drivers/sensor/adxl375/Kconfig:
```diff
@@ -0,0 +1,13 @@
+
+# Copyright (c) 2023 Aaron Chan
+# SPDX-License-Identifier: Apache-2.0
+
```
where is the trigger support option in the Kconfig?

amiclaus reviewed (changes requested):
please properly arrange the commits:
- one for the bindings,
- one for the driver implementation
- one for the tests.

AarC10 reviewed (commented):
Thanks for the feedback. Currently working on addressing all of these comments and will push changes soon.

EDIT: Everything should be pushed now. Added to the build_all tests. Removed trigger code that wasn't completed. Added the adxl375-spi bindings. Addressed any other suggestions for improvement for code. Redid the commits to be separated under bindings, implementation and tests

AarC10 committed:
drivers: sensor: adi: ADXL375 Devicetree Bindings

Define devicetree bindings for ADXL375 to support setting ODR and LP

Signed-off-by: Aaron Chan <aarchan108@gmail.com>

AarC10 committed:
drivers: sensor: adi: ADXL375 Accelerometer Driver Implementation

Implement basic I2C and SPI support for the ADXL375 chip capable of reading acceleration data and setting ODR and low power rate

Signed-off-by: Aaron Chan <aarchan108@gmail.com>

AarC10 committed:
drivers: sensor: adi: ADXL375 Driver build_all tests

Add ADXL375 driver to the build_all tests for confirming successful compile

Signed-off-by: Aaron Chan <aarchan108@gmail.com>
